web/src/js/ducks/flows/index.ts

```diff
@@ -0,0 +1,363 @@
+                flow,
+                makeSort(sort),
+            ));
+        }
+        // Update selection
+        if (selectedIds.has(flow.id)) {
+            selected = selected.map((existing) =>
+                existing.id === flow.id ? flow : existing,
+            );
+        }
+        // Update highlight
+        if (matching_filters[FilterName.Highlight]) {
+            if (!highlighted.has(flow.id)) {
+                highlighted = new Set(highlighted);
+                highlighted.add(flow.id);
+            }
+        } else {
+            highlighted = withElemRemoved(highlighted, flow.id);
+        }
+
+        return {
+            list,
+            _listIndex,
+            byId,
+            view,
+            _viewIndex,
+            sort,
+            selected,
+            selectedIds,
+            highlighted,
+        };
+    } else if (FLOWS_REMOVE.match(action)) {
+        const flow_id = action.payload;
+        const { sort } = state;
+        let { view, _viewIndex, selected, selectedIds } = state;
+        const listPos = state._listIndex.get(flow_id)!;
+        const list = state.list.toSpliced(listPos, 1);
+        const _listIndex = withKeyRemoved(state._listIndex, flow_id);
+        const byId = withKeyRemoved(state.byId, flow_id);
+        // Update view
+        const viewPos = _viewIndex.get(flow_id);
+        if (viewPos !== undefined) {
+            ({ view, _viewIndex } = removeViewItemAt(
+                view,
+                _viewIndex,
+                viewPos,
+            ));
+        }
+        // Update selection
+        if (selectedIds.has(flow_id)) {
+            if (selected.length === 1 && viewPos !== undefined) {
+                const fallback =
+                    view[viewPos /* no +1, already removed */] ??
+                    view[viewPos - 1];
+                selected = [fallback];
+            } else {
+                selected = selected.filter((f) => f.id !== flow_id);
+            }
+            selectedIds = buildLookup(selected);
+        }
+        // Update highlight
+        const highlighted = withElemRemoved(state.highlighted, flow_id);
+
+        return {
+            list,
+            _listIndex,
+            byId,
+            view,
+            _viewIndex,
+            sort,
+            selected,
+            selectedIds,
+            highlighted,
+        };
+    } else if (FLOWS_FILTER_UPDATE.match(action)) {
+        const { name, matching_flow_ids } = action.payload;
+        switch (name) {
+            case FilterName.Search: {
+                const view = matching_flow_ids
+                    .map((id) => state.byId.get(id))
+                    .filter((f) => f !== undefined)
+                    .toSorted(makeSort(state.sort));
+                const _viewIndex = buildIndex(view);
+                return {
+                    ...state,
+                    view,
+                    _viewIndex,
+                };
+            }
```

Comment:
We have a bug here: if we apply a filter and then clear it, no flows are shown because `matching_flow_ids` is set to `[]`. Previously, I think I was also passing the filter expression, and if it was empty, it meant that `matching_flow_ids` could be ignored. Otherwise, we could pass `None` or `undefined` for the `matching_flow_ids`.

Reply:
Great catch, fixed!